### PR TITLE
Contextual evaluations (for other warehouses, excl BQ)

### PR DESCRIPTION
**Pull request by Ekrekr**

### CODE REVIEW

File: api/dbadapters/index.ts [hunk 1]
```diff
@@ -66,3 +68,74 @@ register("postgres", RedshiftDbAdapter);
...
+        );
+        if (concatenate) {
+          validationQueries.push({
+            query: concatenateQueries(tableQueries, queryModifier)
+          });
+        } else {
+          tableQueries.forEach(query => validationQueries.push({ query: queryModifier(query) }));
+        }
+      } else if (queryOrAction instanceof dataform.Operation) {
+        // TODO: The prefix method is a bit sketchy for operations. For example, after attaching the `eplain` prefix
```
Comment (Ekrekr) [inline api/dbadapters/index.ts:114]: This issue already exists with the current method of evaluating operations, so is not a regression

File: api/dbadapters/sqldatawarehouse.ts [hunk 2]
```diff
@@ -90,25 +95,30 @@ export class SQLDataWarehouseDBAdapter implements IDbAdapter {
...
-      await this.execute(`explain ${queryOrAction}`);
-      return [
-        dataform.QueryEvaluation.create({
-          status: dataform.QueryEvaluation.QueryEvaluationStatus.SUCCESS
-        })
-      ];
-    } catch (e) {
-      return [
-        dataform.QueryEvaluation.create({
+    // TODO: Using `explain` before declaring a variable is not valid in SQL Data Warehouse.
```
Comment (Ekrekr) [inline api/dbadapters/sqldatawarehouse.ts:98]: It seems dry runs with variable declarations don't work. Know of any workarounds?

File: api/dbadapters/index.ts (hunk 1 above)
Comment (BenBirt) [inline api/dbadapters/index.ts:114]:
```suggestion
        // TODO: The prefix method is a bit sketchy for operations. For example, after attaching the `explain` prefix
```

Comment (BenBirt) [reply api/dbadapters/index.ts:114]: does this not happen for all action types, though? e.g. if i have a `table` action that looks like this:

```
// some comment
select 1
```

File: api/dbadapters/index.ts [hunk 3]
```diff
@@ -66,3 +68,74 @@ register("postgres", RedshiftDbAdapter);
...
+      } else if (queryOrAction instanceof dataform.Assertion) {
+        validationQueries.push({ query: queryModifier(queryOrAction.query) });
+      } else {
+        throw new Error("Unrecognized evaluate type.");
+      }
+    } catch (e) {
+      throw new ErrorWithCause(`Error building tasks for evaluation. ${e.message}`, e);
+    }
+  }
+  return validationQueries.filter(({ query }) => !!query);
```
Comment (BenBirt) [inline api/dbadapters/index.ts:140]:
```suggestion
  return validationQueries.map(query => query.trim()).filter(({ query }) => !!query);
```

File: api/dbadapters/sqldatawarehouse.ts (hunk 2 above)
Comment (BenBirt) [reply api/dbadapters/sqldatawarehouse.ts:98]: I do not, but I wouldn't worry much about this edge case especially as we have extremely few Azure users

File: api/dbadapters/index.ts [hunk 4]
```diff
@@ -66,3 +68,74 @@ register("postgres", RedshiftDbAdapter);
 register("redshift", RedshiftDbAdapter);
 register("snowflake", SnowflakeDbAdapter);
 register("sqldatawarehouse", SQLDataWarehouseDBAdapter);
+
+export function constructEvaluationFromQueryOrAction(
```
Comment (BenBirt) [inline api/dbadapters/index.ts:72]:
```suggestion
export function collectEvaluationQueries(
```

**Comment by Ekrekr** [comment]
> does this not happen for all action types, though?

Hmm, nice catch. Have moved comment to top of block and changed it a bit.

**Pull request merged**